linalg/kernels.hpp

```diff
@@ -333,28 +333,47 @@ void Set(const int height, const int width,
    }
 }
 
-/** @brief Matrix-matrix multiplication: A = B * C, where the matrices A, B and
-    C are of sizes @a Aheight x @a Awidth, @a Aheight x @a Bwidth and @a Bwidth
-    x @a Awidth, respectively. */
+/** @brief Matrix-matrix multiplication: A = alpha * B * C + beta * A, where the
+    matrices A, B and C are of sizes @a Aheight x @a Awidth, @a Aheight x @a
+    Bwidth and @a Bwidth x @a Awidth, respectively. */
 template<typename TA, typename TB, typename TC>
 MFEM_HOST_DEVICE inline
-void Mult(const int Aheight, const int Awidth, const int Bwidth,
-          const TB *Bdata, const TC *Cdata, TA *Adata)
+void AddMult(const int Aheight, const int Awidth, const int Bwidth,
+             const TB *Bdata, const TC *Cdata, TA *Adata, const TB alpha,
+             const TA beta)
 {
    const int ah_x_aw = Aheight * Awidth;
-   for (int i = 0; i < ah_x_aw; i++) { Adata[i] = 0.0; }
+   if (beta == 0.0)
+   {
+      for (int i = 0; i < ah_x_aw; i++) { Adata[i] = 0.0; }
+   }
+   else if (beta != 1.0)
+   {
+      for (int i = 0; i < ah_x_aw; i++) { Adata[i] *= beta; }
+   }
    for (int j = 0; j < Awidth; j++)
    {
       for (int k = 0; k < Bwidth; k++)
       {
          for (int i = 0; i < Aheight; i++)
          {
-            Adata[i+j*Aheight] += Bdata[i+k*Aheight] * Cdata[k+j*Bwidth];
+            Adata[i+j*Aheight] += alpha * Bdata[i+k*Aheight] * Cdata[k+j*Bwidth];
```

Comment:
`alpha` can be factored out of the loop to replace `Aheight` multiplications with just 1 multiplication.

Reply:
Like this?

```cpp
   for (int j = 0; j < Awidth; j++)
   {
      for (int k = 0; k < Bwidth; k++)
      {
         double val = 0.0;
         for (int i = 0; i < Aheight; i++)
         {
            val += Bdata[i+k*Aheight] * Cdata[k+j*Bwidth];
         }
         Adata[i+j*Aheight] += alpha * val;
      }
   }
```

Reply:
Do you mean by rearranging the loops? To me it looks like `Awidth * Aheight` multiplications by `alpha` are required, so `Bwidth` multiplications could be replaced with 1 by making the loop over `k` into the inner-most loop. Is there a different way?

Reply:
Yeah, you are right -- currently the innermost loop is not the `Bwidth` loop and that's the sum that one can factor `alpha` out of.

What we can do instead is factor `alpha * Cdata[k+j*Bwidth]` out of the innermost loop over `i` -- maybe the compiler already will do that but it does not hurt to make it explicit.

Reply:
Done in 23608099383dc0285b5552497360819c8204bd99.